read-fonts/src/tables/postscript/charstring.rs

```diff
@@ -1,203 +1,644 @@
+                                    y2 + self.stack.get_fixed(i + 4)?
+                                } else {
+                                    y2
+                                };
+                                alternate = true;
+                            }
+                            sink.curve_to(x1, y1, x2, y2, x3, y3);
+                            x = x3;
+                            y = y3;
+                            i += 4;
+                        }
+                        self.stack.clear();
+                    }
+                    // Call local or global subroutine
+                    // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L972>
+                    CALLSUBR | CALLGSUBR => {
+                        let subrs_index = if op == CALLSUBR {
+                            self.subrs.as_ref().unwrap()
+                        } else {
+                            &self.global_subrs
+                        };
+                        let biased_index =
+                            (self.stack.pop_i32()? + subrs_index.subr_bias()) as usize;
+                        let subr_charstring_data = subrs_index.get(biased_index)?;
+                        let pos =
+                            self.evaluate(subr_charstring_data, x, y, sink, nesting_depth + 1)?;
+                        x = pos.0;
+                        y = pos.1;
+                    }
+                    _ => return Err(Error::InvalidCharstringOperator(op)),
+                }
+            }
+        }
+        Ok((x, y))
+    }
 }
 
-fn parse_token(cursor: &mut Cursor) -> Result<Token, Error> {
-    // Escape opcode for accessing extensions.
-    const ESCAPE: u8 = 12;
-    let b0 = cursor.read::<u8>()?;
-    Ok(if b0 == ESCAPE {
-        let b1 = cursor.read::<u8>()?;
-        Token::Operator(
-            Operator::from_extended_opcode(b1).ok_or(Error::InvalidCharstringOperator(b1))?,
-        )
-    } else {
-        // See <https://learn.microsoft.com/en-us/typography/opentype/spec/cff2#table-3-operand-encoding>
-        match b0 {
-            // Ints are the same encoding as DICTs
-            28 | 29 | 32..=254 => super::dict::parse_int(cursor, b0)?.into(),
-            // Reals are just 16.16 fixed point values though rather than BCD
-            255 => Fixed::from_bits(cursor.read::<i32>()?).into(),
-            _ => Token::Operator(
-                Operator::from_opcode(b0).ok_or(Error::InvalidCharstringOperator(b0))?,
-            ),
-        }
-    })
+/// Charstring operators.
+/// See <https://learn.microsoft.com/en-us/typography/opentype/spec/cff2charstr#appendix-a-cff2-charstring-command-codes>
+// TODO: This is currently missing legacy math and logical operators.
+// fonttools doesn't even implement these: <https://github.com/fonttools/fonttools/blob/65598197c8afd415781f6667a7fb647c2c987fff/Lib/fontTools/misc/psCharStrings.py#L409>
+mod ops {
```

Comment:
I would probably replace this with an enum that could represent both the one-byte and two-byte operators, and then have a little convenience method that parses that out, but this works too.

Reply:
+1

Reply:
Done